torch/quantization/observer.py

```diff
@@ -195,27 +165,27 @@ def _calculate_qparams(self, min_val, max_val):
             else:
                 qmin, qmax = 0, 255
 
-        max_val, min_val = float(max_val), float(min_val)
-        min_val = min(0.0, min_val)
-        max_val = max(0.0, max_val)
-        if max_val == min_val:
-            scale = 1.0
-            zero_point = 0
+        min_val = torch.min(min_val, torch.zeros_like(min_val))
+        max_val = torch.max(max_val, torch.zeros_like(max_val))
+
+        scale = torch.ones(min_val.size(), dtype=torch.float32)
+        zero_point = torch.zeros(min_val.size(), dtype=torch.int64)
+        device = 'cuda' if min_val.is_cuda else 'cpu'
+
+        if self.qscheme == torch.per_tensor_symmetric or self.qscheme == torch.per_channel_symmetric:
+            max_val = torch.max(-min_val, max_val)
+            scale = max_val / (float(qmax - qmin) / 2)
+            scale = torch.max(scale, torch.tensor(1.0, device=device).new_full(scale.size(), self.eps, dtype=scale.dtype))
+            if self.dtype == torch.quint8:
+                zero_point = zero_point.new_full(zero_point.size(), 128)
         else:
-            if self.qscheme == torch.per_tensor_symmetric or self.qscheme == torch.per_channel_symmetric:
-                max_val = max(-min_val, max_val)
-                scale = max_val / ((qmax - qmin) / 2)
-                scale = max(scale, self.eps)
-                zero_point = 0 if self.dtype == torch.qint8 else 128
-            else:
-                scale = (max_val - min_val) / float(qmax - qmin)
-                scale = max(scale, self.eps)
-                zero_point = qmin - round(min_val / scale)
-                zero_point = max(qmin, zero_point)
-                zero_point = min(qmax, zero_point)
-                zero_point = int(zero_point)
+            scale = (max_val - min_val) / float(qmax - qmin)
+            scale = torch.max(scale, torch.tensor(1.0, device=device).new_full(scale.size(), self.eps, dtype=scale.dtype))
+            zero_point = qmin - torch.round(min_val / scale)
+            zero_point = torch.max(zero_point, torch.tensor(1.0, device=device).new_full(zero_point.size(), qmin, dtype=zero_point.dtype))
```

Comment:
Zero-point should be in the range of qmin and qmax.  Cleaner way to do this would be (broadcasting is supported)
```python
zero_point  = torch.max(zero_point, torch.tensor(qmin, dtype=zero_point.dtype))
```